twake/frontend/src/app/scenes/Client/Popup/AddUser/MagicLinks.tsx

```diff
@@ -0,0 +1,120 @@
+import React, { useEffect, useState } from 'react';
+import { Button, Input, Space, message } from 'antd';
+import { DeleteOutlined, RetweetOutlined } from '@ant-design/icons';
+import Languages from 'services/languages/languages';
+import RouterServices from 'services/RouterService';
+import './MagicLinks.scss';
+import Api from 'app/services/Api';
+
+
+type PropsType = {
+    [key: string]: any;
+};
+
+type MagicLinksResponse = {
+    token: string
+}
+
+class MagicLinksService {
+
+    constructor(protected companyId: string, protected workspaceId: string, protected loading = (arg: boolean) => { }) { }
+
+    private route = `/internal/services/workspaces/v1/companies/${this.companyId}/workspaces/${this.workspaceId}/users/tokens`;
+
+    getCurrentTokens(): Promise<MagicLinksResponse[] | null> {
+        this.loading(true);
+        return Api.get<{ resources: MagicLinksResponse[] }>(this.route).then((a) => (a.resources) ? a.resources : null).finally(() => this.loading(false));
+    }
+
+    recreateToken(): Promise<MagicLinksResponse> {
+        this.loading(true);
+        return Api.post<any, { resource: MagicLinksResponse }>(this.route, {}).then(a => a.resource).finally(() => this.loading(false));
+    }
+
+    deleteToken(token: string): Promise<undefined> {
+        this.loading(true);
+        return Api.delete(this.route + '/' + token).then(a => undefined).finally(() => this.loading(false));
+    }
+}
```

Comment:
My only feedback, this class should be moved in a MagicLinksApiClient.ts file in services/workspaces
Also our convention is to use
```
`${this.route}/${…}`
```
Instead of string concatenation with « + »

Reply:
If you allow me to disagree — I would do this! )
The one of the coolest things of react — it's a component isolation. It's ok to have services outside of component is the case of multiple usage. But data related to one component and used only on one component is even better to keep inside of this component (as far as framework supports and approve this).

But last word is yours, anyway.

Concatenation fix coming soon.

Reply:
I agree, it means that we have API logic both in components and in services depending on feature.
How can we avoid this ? There is some cleanness to have all the hooks (useSomeyhing) in the same folder for reference for instance.
Just take this in consideration and do what you prefer, it is no big work to change in the future if we want to refactor this anyway :)